common/lockmanager_test.go

```diff
@@ -0,0 +1,318 @@
+    }
+
+    // Wait for the thread to finish unlocking key2.
+    threadWaitGroup.Wait();
+}
+
+func TestLockUnlockDifferentKeysTwoThreads(test *testing.T) {
+    key1 := "testkey1";
+    key2 := "testkey2";
+    var threadsWaitGroup sync.WaitGroup;
+    var funcWaitGroup sync.WaitGroup
+
+    threadsWaitGroup.Add(2);
+    funcWaitGroup.Add(2);
+
+    // First thread locks and unlocks key1.
+    go func() {
+        defer funcWaitGroup.Done();
+        Lock(key1);
+        threadsWaitGroup.Done();
+        err := Unlock(key1);
+        if (err != nil) {
+            test.Fatalf("Failed to unlock.")
+        }
+    }()
+
+    // Second thread locks and unlocks key2.
+    go func() {
+        defer funcWaitGroup.Done();
+        Lock(key2);
+        threadsWaitGroup.Done();
+        err := Unlock(key2);
+        if (err != nil) {
+            test.Fatalf("Failed to Unlock")
+        }
+    }()
+    
+    // Wait until both threads have finished locking their keys so they 
+    // can unlock at the same time.
+    threadsWaitGroup.Wait();
+    
+    // Wait until both threads finish their execution.
+    funcWaitGroup.Wait();
+}
+
+func TestLockingTwiceWithSameKeyConcurrently(test *testing.T) {
+    key1 := "testkey1";
+    var firstThreadWaitGroup sync.WaitGroup;
+    var secondThreadWaitGroup sync.WaitGroup;
+    var bothThreadsWaitGroup sync.WaitGroup;
+
+    firstThreadWaitGroup.Add(1);
+    secondThreadWaitGroup.Add(1);
+    bothThreadsWaitGroup.Add(1);
+
+    // First thread aquires the lock a first time.
+    go func() {
+        defer firstThreadWaitGroup.Done();
+        Lock(key1);
+    }()
+
+    // Wait for the first thread to lock key1. 
+    firstThreadWaitGroup.Wait();
+
+
+    // Second thread aquires the lock a second time.
+    go func() {
+        defer bothThreadsWaitGroup.Done();
+        secondThreadWaitGroup.Done();
+        Lock(key1);
+    }()
+
+    // Wait for the second thread to try to lock key1
+    // while the lock is being used by the first thread.
+    secondThreadWaitGroup.Wait();
+    
+    // Small sleep to ensure the second thread gets to locking key1.
+    time.Sleep(10 * time.Millisecond);
+
+    // Unlock the first threads lock.
+    err := Unlock(key1);
+    if (err != nil) {
+        test.Fatalf("Failed to unlock.");
+    }
+    
+    // Wait for the second thread to aquire the lock.
+    bothThreadsWaitGroup.Wait();
+
+    // Unlock the second threads lock.
+    err = Unlock(key1);
+    if (err != nil) {
+        test.Fatalf("Failed to unlock.");
+    }
+}
+
+func TestLockUnlockSameKeyConcurrently(test *testing.T) {
+    key1 := "testkey1";
+    var threadWaitGroup sync.WaitGroup;
+    var verifyWaitGroup sync.WaitGroup;
+
+    // Lock key1 for the first time.
+    Lock(key1);
+
+    threadWaitGroup.Add(1);
+    verifyWaitGroup.Add(1);
+
+    // This thread should wait to lock/unlock key1 a second time until it gets
+    // unlocked for the first time.
+    go func() {
```

Comment:
The double defer pattern in this routine is very likely to confuse people (especially if they don't have a good grasp on defer ordering).
Do we need the second defer?
It will happen before the first one anyways, why not just call it directly?